L1Trigger/TrackFindingTracklet/src/DuplicateRemoval.cc

```diff
@@ -0,0 +1,142 @@
+#include "L1Trigger/TrackFindingTracklet/interface/DuplicateRemoval.h"
+
+#include <vector>
+#include <numeric>
+#include <algorithm>
+
+using namespace std;
+using namespace edm;
+using namespace tt;
+using namespace trackerTFP;
+
+namespace trklet {
+
+  DuplicateRemoval::DuplicateRemoval(const ParameterSet& iConfig,
+                                     const Setup* setup,
+                                     const DataFormats* dataFormats,
+                                     const ChannelAssignment* channelAssignment,
+                                     int region)
+      : enableTruncation_(iConfig.getParameter<bool>("EnableTruncation")),
+        setup_(setup),
+        dataFormats_(dataFormats),
+        channelAssignment_(channelAssignment),
+        region_(region) {}
+
+  // read in and organize input tracks and stubs
+  void DuplicateRemoval::consume(const StreamsTrack& streamsTrack, const StreamsStub& streamsStub) {
+    auto nonNullTrack = [](int& sum, const FrameTrack& frame) { return sum += (frame.first.isNonnull() ? 1 : 0); };
+    auto nonNullStub = [](int& sum, const FrameStub& frame) { return sum += (frame.first.isNonnull() ? 1 : 0); };
+    // count tracks and stubs and reserve corresponding vectors
+    int sizeStubs(0);
+    const int offset = region_ * setup_->numLayers();
+    const StreamTrack& streamTrack = streamsTrack[region_];
+    input_.reserve(streamTrack.size());
+    const int sizeTracks = accumulate(streamTrack.begin(), streamTrack.end(), 0, nonNullTrack);
+    for (int layer = 0; layer < setup_->numLayers(); layer++) {
+      const StreamStub& streamStub = streamsStub[offset + layer];
+      sizeStubs += accumulate(streamStub.begin(), streamStub.end(), 0, nonNullStub);
+    }
+    tracks_.reserve(sizeTracks);
+    stubs_.reserve(sizeStubs);
+    // transform input data into handy structs
+    for (int frame = 0; frame < (int)streamTrack.size(); frame++) {
+      const FrameTrack& frameTrack = streamTrack[frame];
+      if (frameTrack.first.isNull()) {
+        input_.push_back(nullptr);
+        continue;
+      }
+      vector<Stub*> stubs;
```

Comment:
Between L48-L78, it's not at all clear what is going on. To save people having to understand every line of code here, please add comment explaining it. You already had streamTrack & streamStub. What is the purpose of creating new FrameStub objects etc. and storing them in your structs?

Reply:
done